File: cliente.cpp
#include <iostream>
#include <string>
#include <grpcpp/grpcpp.h>
#include "notas.grpc.pb.h"

class NotasClient {
public:
    NotasClient(std::shared_ptr<grpc::Channel> channel)
        : stub_(gerencia_notas::GerenciadorNotas::NewStub(channel)) {}


    void AdicionarNotaExemplo(const std::string& ra, const std::string& cod_disciplina, double nota) {
        gerencia_notas::AdicionaNotaRequest request;
        request.set_ra(ra);
        request.set_cod_disciplina(cod_disciplina);
        request.set_nota(nota);

        gerencia_notas::StatusResponse reply;
        grpc::ClientContext context;

        grpc::Status status = stub_->AdicionarNota(&context, request, &reply);

        std::cout << "[AdicionarNota] ";
        if (status.ok()) {
            std::cout << reply.msg() << std::endl;
        } else {
            std::cout << "Erro RPC: " << status.error_code()
                      << ": " << status.error_message() << std::endl;
        }
    }

    void AlterarNotaExemplo(const std::string& ra, const std::string& cod_disciplina, double nova_nota) {
        gerencia_notas::AdicionaNotaRequest request;
        request.set_ra(ra);
        request.set_cod_disciplina(cod_disciplina);
        request.set_nota(nova_nota);

        gerencia_notas::StatusResponse reply;
        grpc::ClientContext context;

        grpc::Status status = stub_->AlterarNota(&context, request, &reply);

        std::cout << "[AlterarNota] ";
        if (status.ok()) {
            std::cout << reply.msg() << std::endl;
        } else {
            std::cout << "Erro RPC: " << status.error_code()
                      << ": " << status.error_message() << std::endl;
        }
    }

    void ConsultarNotaExemplo(const std::string& ra, const std::string& cod_disciplina) {
        gerencia_notas::AlunoDisciplinaRequest request;
        request.set_ra(ra);
        request.set_cod_disciplina(cod_disciplina);

        gerencia_notas::ConsultaNotaResponse reply;
        grpc::ClientContext context;

            grpc::Status status = stub_->ConsultarNota(&context, request, &reply);

        std::cout << "[ConsultarNota] ";
        if (status.ok()) {
                if (reply.sucesso()) {
                    std::cout << "Nota consultada com sucesso." << std::endl;
                } else {
                    std::cout << reply.msg_erro() << std::endl;
                }
        } else {
            std::cout << "Erro RPC: " << status.error_code()
                      << ": " << status.error_message() << std::endl;
        }
    }

    void CalcularMediaExemplo(const std::string& ra) {
        gerencia_notas::AlunoRequest request;
        request.set_ra(ra);

        gerencia_notas::MediaResponse reply;
        grpc::ClientContext context;

        grpc::Status status = stub_->CalcularMedia(&context, request, &reply);

        std::cout << "[CalcularMedia] ";
        if (status.ok()) {
            if (reply.sucesso()) {
                std::cout << "Média: " << reply.media() << std::endl;
            } else {
                std::cout << reply.msg_erro() << std::endl;
            }
        } else {
            std::cout << "Erro RPC: " << status.error_code()
                      << ": " << status.error_message() << std::endl;
        }
    }

    void ListarNotasAlunoExemplo(const std::string& ra) {
        gerencia_notas::AlunoRequest request;
        request.set_ra(ra);

        grpc::ClientContext context;
        std::unique_ptr<grpc::ClientReader<gerencia_notas::Nota>> reader(
            stub_->ListarNotasAluno(&context, request));
        gerencia_notas::Nota nota;
        std::cout << "[ListarNotasAluno] Notas do RA " << ra << ":\n";
        bool recebeu = false;
        while (reader->Read(&nota)) {
            recebeu = true;
            std::cout << "  Disciplina: " << nota.cod_disciplina()
                      << ", Nota: " << nota.nota() << std::endl;
        }
        grpc::Status status = reader->Finish();
        if (!recebeu) {
            std::cout << "  Nenhuma nota recebida ou funcionalidade não implementada." << std::endl;
        }
        if (!status.ok()) {
            std::cout << "  Status: " << status.error_message() << std::endl;
        }
    }

    /*S
    void ListarNotasExemplo(const std::string& ra) {
        // TODO: implementar streaming
    }
    */

private:
    std::unique_ptr<gerencia_notas::GerenciadorNotas::Stub> stub_;
};

int main(int argc, char** argv) {
    NotasClient client(grpc::CreateChannel("localhost:50052",
                                           grpc::InsecureChannelCredentials()));

    std::string ra = "456";

    // Adicionando várias notas para diferentes disciplinas
    client.AdicionarNotaExemplo(ra, "CIC0002", 8.0);
    client.AdicionarNotaExemplo(ra, "CIC0003", 7.5);
    client.AdicionarNotaExemplo(ra, "CIC0004", 9.2);
    client.AdicionarNotaExemplo(ra, "CIC0005", 6.8);

    // Alterando uma nota
    client.AlterarNotaExemplo(ra, "CIC0002", 9.5);

    // Consultando uma nota específica
    client.ConsultarNotaExemplo(ra, "CIC0002");

    // Calculando média
    client.CalcularMediaExemplo(ra);

    // Listando todas as notas do aluno
    client.ListarNotasAlunoExemplo(ra);

    return 0;
}

File: python/cliente.py
# cliente.py (esqueleto)
import grpc
import notas_pb2
import notas_pb2_grpc

def run():
    with grpc.insecure_channel('localhost:50051') as channel:
        stub = notas_pb2_grpc.GerenciadorNotasStub(channel)



        ra = "123"
        # Adicionando várias notas para diferentes disciplinas
        disciplinas_notas = [
            ("CIC0001", 9.5),
            ("CIC0002", 8.7),
            ("CIC0003", 7.8),
            ("CIC0004", 6.9)
        ]
        for cod_disciplina, nota in disciplinas_notas:
            response = stub.AdicionarNota(notas_pb2.AdicionaNotaRequest(
                ra=ra,
                cod_disciplina=cod_disciplina,
                ano=2025,
                semestre=2,
                nota=nota
            ))
            print(f"Resposta do servidor (AdicionarNota {cod_disciplina}): {response.msg}")

        # Alterando uma nota
        response = stub.AlterarNota(notas_pb2.AdicionaNotaRequest(
            ra=ra,
            cod_disciplina="CIC0001",
            ano=2025,
            semestre=2,
            nota=8.0
        ))
        print(f"Resposta do servidor (AlterarNota CIC0001): {response.msg}")

        # Consultando uma nota específica
        consulta_response = stub.ConsultarNota(notas_pb2.AlunoDisciplinaRequest(
            ra=ra,
            cod_disciplina="CIC0001"
        ))
        if consulta_response.sucesso:
            nota = consulta_response.nota
            print(f"Nota consultada: RA={nota.ra}, Disciplina={nota.cod_disciplina}, Ano={nota.ano}, Semestre={nota.semestre}, Nota={nota.nota}")
        else:
            print(f"Erro ao consultar nota: {consulta_response.msg_erro}")

        # Calcular média
        media_response = stub.CalcularMedia(notas_pb2.AlunoRequest(ra=ra))
        if media_response.sucesso:
            print(f"Média do aluno {ra}: {media_response.media}")
        else:
            print(f"Erro ao calcular média: {media_response.msg_erro}")

        # --- CHAMADA DO DESAFIO (STREAMING) ---
        # TODO: Descomente este bloco de código após implementar o desafio 
        # no servidor.
        # O objetivo é iterar sobre o stream de respostas recebido do 
        # servidor e imprimir cada nota.
        
        print("\n--- Listando todas as notas do aluno 123 (DESAFIO) ---")
        try:
            notas_stream = stub.ListarNotasAluno(notas_pb2.AlunoRequest(ra="123"))
            for nota in notas_stream:
                print(f"  - Disciplina: {nota.cod_disciplina}, Nota: {nota.nota}")
        except grpc.RpcError as e:
            print(f"Erro ao chamar ListarNotasAluno: {e.details()}")
        

if __name__ == '__main__':
    run()